casper/src/main/resources/TreeHashMap.rho

```diff
@@ -0,0 +1,220 @@
+        }
+      } |
+
+      contract ByteArrayToNybbleList(@ba, @n, @len, @acc, ret) = {
+        if (n == len) {
+          ret!(acc)
+        } else {
+          ByteArrayToNybbleList!(ba, n+1, len, acc ++ [ ba.nth(n) % 16, ba.nth(n) / 16 ], *ret)
+        }
+      } |
+
+      contract TreeHashMap(@"init", @depth, ret) = {
+        new map in {
+          MakeNode!(0, (*map, [])) |
+          @(*map, "depth")!!(depth) |
+          ret!(*map)
+        }
+      } |
+
+      contract TreeHashMapGetter(@map, @nybList, @n, @len, @suffix, ret) = {
+        // Look up the value of the node at [map, nybList.slice(0, n + 1)]
+        new valCh in {
+          nodeGet!((map, nybList.slice(0, n)), *valCh) |
+          for (@val <- valCh) {
+            if (n == len) {
+              ret!(val.get(suffix))
+            } else {
+              // Otherwise check if the rest of the path exists.
+              // Bit k set means node k exists.
+              // nybList.nth(n) is the node number
+              // val & powers.nth(nybList.nth(n)) is nonzero if the node exists
+              // (val/powers.nth(nybList.nth(n))) % 2 is 1 if the node exists
+              if ((val/powers.nth(nybList.nth(n))) % 2 == 0) {
+                ret!(Nil)
+              } else {
+                TreeHashMapGetter!(map, nybList, n + 1, len, suffix, *ret)
+              }
+            }
+          }
+        }
+      } |
+
+      contract TreeHashMap(@"get", @map, @key, ret) = {
+        new hashCh, nybListCh in {
+          // Hash the key to get a 256-bit array
+          @"keccak256Hash"!(key.toByteArray(), *hashCh) |
+          for (@hash <- hashCh; @depth <- @(map, "depth")) {
+            // Get the bit list
+            ByteArrayToNybbleList!(hash, 0, depth, [], *nybListCh) |
+            for (@nybList <- nybListCh) {
+              TreeHashMapGetter!(map, nybList, 0, 2 * depth, hash.slice(depth, 32), *ret)
+            }
+          }
+        }
+      } |
+
+      // Doesn't walk the path, just tries to fetch it directly.
+      // Will hang if there's no key with that 64-bit prefix.
+      // Returns [] like "get" does if there is some other key with
+      // the same prefix but no value there.
+      contract TreeHashMap(@"fastUnsafeGet", @map, @key, ret) = {
+        new hashCh, nybListCh in {
+          // Hash the key to get a 256-bit array
+          @"keccak256Hash"!(key.toByteArray(), *hashCh) |
+          for (@hash <- hashCh; @depth <- @(map, "depth")) {
+            // Get the bit list
+            ByteArrayToNybbleList!(hash, 0, depth, [], *nybListCh) |
+            for (@nybList <- nybListCh) {
+              new restCh, valCh in {
+                nodeGet!((map, nybList), *restCh) |
+                for (@rest <- restCh) {
+                  valCh!(rest.get(hash.slice(depth, 32))) |
+                  for (@val <- valCh) {
+                    if (val == Nil) { ret!(Nil) }
+                    else { ret!(val) }
```

Comment:
I think this is equivalent to `ret!(rest.get(hash.slice(depth, 32)))`?

Reply:
Oops, yes.  Thx